zephyr/lib/alloc.c

```diff
@@ -212,15 +236,24 @@ static void heap_free(struct k_heap *h, void *mem)
 
 	if (is_cached(mem)) {
 		mem_uncached = z_soc_uncached_ptr((__sparse_force void __sparse_cache *)mem);
+#if CONFIG_L3_HEAP
+		if (h == &l3_heap)
+			mem_uncached = mem;
```

Comment:
this looks so strange... We take great care in this function to separate cached and uncached pointers and this just mixes them up. Maybe we need to introduce a new type of "cached heap?"

Reply:
fixed